xunit-autolabeler-v2/ast_parser/lib/file_utils.py

```diff
@@ -0,0 +1,116 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import List, Callable
+
+import re
+import os
+import subprocess
+
+from . import constants
+
+
+def _getFiles(root_dir: str, predicate: Callable[[str], bool]) -> List[str]:
+    """Recursively list the files in a given directory
+       whose names match the provided predicate function
+
+    Args:
+        root_dir: the root directory to search from
+        predicate: the predicate function
+                                       to filter filenames with
+
+    Returns:
+        A list of filepaths relative to root_dir that match the predicate
+    """
+    if os.path.basename(os.path.normpath(root_dir)).startswith('.'):
+        # Ignore dot-directories
+        return []
+
+    paths = [os.path.join(root_dir, path) for path in os.listdir(root_dir)]
+
+    folders = [path for path in paths if not os.path.isfile(path)]
+
+    files = [path for path in paths if os.path.isfile(path)
+             and predicate(path)]
+
+    for file in folders:
+        files += _getFiles(file, predicate)
+
+    return files
+
+
+def get_python_files(root_dir: str) -> List[str]:
+    """Recursively lists the Python files in a directory
+
+    Args:
+        root_dir: the root directory to search from
+
+    Returns:
+        A list of Python filepaths relative to root_dir
+    """
+
+    # Not language-agnostic, so keep it in this method
+    gae_lib_regex = re.compile('/appengine/(.+/)*lib/')
+
+    return _getFiles(
+        root_dir,
+        lambda path: (
+            path.endswith('.py') and not gae_lib_regex.search(path)
+        )
+    )
+
+
+def get_drift_yaml_files(root_dir: str) -> List[str]:
+    """Recursively lists the DRIFT yaml metadata files in a directory
+
+    Args:
+        root_dir: the root directory to search from
+
+    Returns:
+        A list of DRIFT yaml metadata filepaths relative to root_dir
+    """
+    return _getFiles(
+        root_dir,
+        lambda path: (
+            os.path.basename(path) == '.drift-data.yml'
+            or os.path.basename(path) == '.drift-data.yaml'
+        )
+    )
+
+
+def get_region_tags(root_dir: str) -> List[str]:
+    """Recursively find the region tags in a directory using `grep`
+
+    Args:
+        root_dir: the root directory to search from
+
+    Returns:
+        The list of region tags found in root_dir
+    """
+    proc = subprocess.Popen(
+        constants.REGION_TAG_GREP_ARGS,
+        stdout=subprocess.PIPE,
+        cwd=root_dir)
+    region_tags = proc.stdout.read().decode().split('\n')
+
+    # Extract region tags from START clauses
+    # e.g. "[START some_tag]" --> "some_tag"
+    region_tags = [tag.lstrip('/#*').strip().strip('[]').strip()
```

Comment:
Sorry again. I have some suggestions for making the code here more readable.

Can we have a single regex to extract the tag name?

maybe (sorry I haven't tested, but I hope you'll get the idea):

```
REGION_TAG_REGEX = re.compile(r"[START\s([^\]]+)]")
...
    region_tags = []
    for tag in region_tags:
        m = constants.REGION_TAG_REGEX.match(tag)
        if m:
            region_tags.append(m.group(1))
    return list(set(region_tags))
```

Reply:
IMO, group capturing is a tad more _complex_ than using the stripping approach.

If you prefer _concise_ code instead however, then we can switch to it. But personally, I find most `regex`es harder to understand after the fact. 🙂